Sources/Purchasing/StoreKit2/Observer Mode/StoreKit2ObserverModePurchaseDetector.swift

```diff
@@ -0,0 +1,109 @@
+
+    private let deviceCache: DeviceCache
+    private let currentUserProvider: CurrentUserProvider
+    private let allTransactionsProvider: AllTransactionsProviderType
+
+    init(
+        deviceCache: DeviceCache,
+        currentUserProvider: CurrentUserProvider,
+        allTransactionsProvider: AllTransactionsProviderType
+    ) {
+        self.deviceCache = deviceCache
+        self.currentUserProvider = currentUserProvider
+        self.allTransactionsProvider = allTransactionsProvider
+    }
+
+    /// Detects unobserved transactions and forwards them to the StoreKit2ObserverModeManagerDelegate
+    /// for processing.
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModeManagerDelegate?
+    ) async {
+        guard let mostRecentVerifiedTransaction: (
+            verifiedTransaction: StoreKit.Transaction,
+            jwsRepresentation: String
+        ) = (await allTransactionsProvider.getAllTransactions()
+            .compactMap { transaction in
+                guard let verifiedTransaction = transaction.verifiedTransaction else {
+                    return nil
+                }
+                return (
+                    verifiedTransaction: verifiedTransaction,
+                    jwsRepresentation: transaction.jwsRepresentation
+                )
+            }
+            .sorted(by: { $0.verifiedTransaction.purchaseDate > $1.verifiedTransaction.purchaseDate })
+            .first
+        ) else {
+            return
+        }
+
+        let transaction: StoreKit.Transaction = mostRecentVerifiedTransaction.verifiedTransaction
+        let jwsRepresentation: String = mostRecentVerifiedTransaction.jwsRepresentation
+
+        var cachedSyncedSK2TransactionIDs = Set(
+            self.deviceCache.cachedSyncedSK2TransactionIDs(appUserID: currentUserProvider.currentAppUserID) ?? []
+        )
+
+        guard !cachedSyncedSK2TransactionIDs.contains(transaction.id) else {
+            return
+        }
+
+        do {
+            try await delegate?.handleSK2ObserverModeTransaction(
+                verifiedTransaction: transaction,
+                jwsRepresentation: jwsRepresentation
+            )
+
+            cachedSyncedSK2TransactionIDs.insert(transaction.id)
+            self.deviceCache.cacheSyncedSK2TransactionIDs(
+                syncedSK2TransactionIDs: Array(cachedSyncedSK2TransactionIDs),
+                appUserID: currentUserProvider.currentAppUserID
+            )
+        } catch {
+            Logger.error(Strings.purchase.sk2_observer_mode_error_processing_transaction(error))
+        }
+    }
+}
+
+/// A wrapper protocol that allows for abstracting out calls to an `AsyncSequence<VerificationResult<Transaction>>`.
+/// This will usually be `Transaction.all` in production but allows us to inject custom AsyncSequences for testing.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+protocol AllTransactionsProviderType: Sendable {
```

Comment:
I initially tried to inject `Transaction.all` into the `StoreKit2ObserverModePurchaseDetector` directly into the actor's initializer (similar to [how it's done in StoreKit2TransactionLister](https://github.com/RevenueCat/purchases-ios/blob/2cff931fa3ddd7f5471cf2bfc641bc6b27651557/Sources/Purchasing/StoreKit2/StoreKit2TransactionListener.swift#L79)), and while this compiled, for some reason it caused `Transaction.all` to not return the initial purchase when the app was foregrounded, so I abandoned that approach.

While it's a bit more verbose, this approach avoids that issue!